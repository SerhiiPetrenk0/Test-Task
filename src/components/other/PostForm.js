import React , { useState } from 'react';
import { Form , Button } from 'react-bootstrap';
import { postCommentAPI } from '../../api';
import { colors } from '../../styled/globalStyled';
import { 
    StyledContainer,
    StyledFormGroup,
    StyledBsStarFill
} from '../../styled/other/PostForm';

export const PostForm = () => {
    const [formValue, setFormValue] = useState(
        {
            id: 'B0721KZEEE',
            rating: '',
            coments: ''
        }
    );
    const [currentValue, setCurrentValue] = useState(0);
    const [hoverValue, setHoverValue] = useState(undefined);
    const stars = Array(5).fill(0);
  
    const handleClick = value => {
      setCurrentValue(value);
      setFormValue({...formValue, rating:value});
    };
  
    const handleMouseOver = newHoverValue => setHoverValue(newHoverValue);

    const handleMouseLeave = () => setHoverValue(undefined);

    const handleTextarea = val => setFormValue({...formValue, coments:val});

    const handleSubmit = event => {
        event.preventDefault();
        postCommentAPI(formValue);
    };

    return(
        <StyledContainer>
            <Form onSubmit={handleSubmit}>
                <StyledFormGroup controlId="exampleForm.ControlInput1">
                    {stars.map((_, index) => (
                            <StyledBsStarFill
                            key={index}
                            size={24}
                            onClick={() => handleClick(index + 1)}
                            onMouseOver={() => handleMouseOver(index + 1)}
                            onMouseLeave={handleMouseLeave}
                            color={(hoverValue || currentValue) > index ? colors.starActive : colors.starDisable}
                            />
                    ))}
                </StyledFormGroup>
                <StyledFormGroup controlId="exampleForm.ControlTextarea1">
                    <Form.Label>comment</Form.Label>
                    <Form.Control as="textarea" rows={3} value={formValue.coments} onChange={event => handleTextarea(event.target.value)} />
                </StyledFormGroup>
                  <Button variant="primary" type="submit">
                        Submit
                  </Button>
            </Form>
        </StyledContainer>
    );
};
