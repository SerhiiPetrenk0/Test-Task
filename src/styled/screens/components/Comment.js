import styled from 'styled-components'
import { Card } from 'react-bootstrap'

export const StyledCard = styled(Card)`
    margin-top: 20px;
    width: 100%
`;