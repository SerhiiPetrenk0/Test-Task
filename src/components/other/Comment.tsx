import React from 'react';
import { Card } from 'react-bootstrap';
import { StarBar } from './StarBar';
import { StyledCard } from '../../styled/other/Comment';

export const Comment = (props: { ListComment: any; }) => {
    const { ListComment } = props;
    return (
        <StyledCard>
            <Card.Body>
                <Card.Text>
                {ListComment.comments}
                </Card.Text>
                <Card.Subtitle>
                <StarBar count={ListComment.rating} />
                </Card.Subtitle>
            </Card.Body>
        </StyledCard>
    );
};
