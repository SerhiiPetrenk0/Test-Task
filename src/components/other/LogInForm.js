import React from "react"
import * as Yup from 'yup'
import { Form, Button, Modal } from "react-bootstrap"
import { useDispatch  } from "react-redux"
import { useFormik } from "formik"
import { StyledValidEmail } from "../../styled/other/LogInForm"
import { fillUser } from "../../redux/redusers/userInfoReduser"

export const LogInForm = props => {
    const { show, handleClose } = props
    const dispatch = useDispatch()
    const formik = useFormik({
      initialValues:{
        email: '',
        password: ''
      },
      validationSchema: Yup.object({
        email: Yup
        .string()
        .max(30, 'Max simbol 15))')
        .required('* this field is required')
        .matches(/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/ , "Invalid email"),
        password: Yup
        .string()
        .required('passwordrequired')
        .matches(/^(?=.*[0-9])(?=.*[!@#$%^&*])[a-zA-Z0-9!@#$%^&*]{6,16}$/ , 'FALSE PASSWORD')
      }),
      onSubmit: values => {
        console.log(values)
        dispatch(fillUser(formik.values))
        handleClose()
      }
    })
    //почистити код / повиносити змінні / проблеми із змінною check / регулярки в змінні / 
      const emaillog = !!formik.errors.email ? formik.errors.email : "We'll never share your email with anyone else."
      const passwordlog = !!formik.errors.password ? formik.errors.password : "Your password must be 6 to 16 characters long, contain letters, numbers, special characters !@#$%^&*, And no spaces or emojis."
      const emailMassage = <StyledValidEmail check={!!formik.errors.email}>{emaillog}</StyledValidEmail>
      const passwordMassage = <StyledValidEmail check={!!formik.errors.password}>{passwordlog}</StyledValidEmail>
     console.log(!!formik.errors.password)
    return (
<>
  <Modal show={show} onHide={handleClose} backdrop="static" keyboard={false}>
    <Modal.Header closeButton>
      <Modal.Title>Modal title</Modal.Title>
    </Modal.Header>
    <Modal.Body>
      <Form onSubmit={formik.handleSubmit}>
        <Form.Group className="mb-3" controlId="formBasicEmail">
          <Form.Label>Email address</Form.Label>
          <Form.Control name="email" onChange={formik.handleChange} value={formik.values.email} type="email" placeholder="Enter email" />
          {emailMassage}
        </Form.Group>

        <Form.Group className="mb-3" controlId="formBasicPassword">
          <Form.Label>Password</Form.Label>
          <Form.Control name="password" onChange={formik.handleChange} value={formik.values.password} type="password" placeholder="Password" />
          {passwordMassage}
        </Form.Group>

        <Button variant="primary" type="submit">
          Submit
        </Button>
      </Form>
    </Modal.Body>
  </Modal>
</>
    )
} 