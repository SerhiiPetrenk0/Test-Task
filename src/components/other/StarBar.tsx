import React from 'react';

import { StyledBsStarFill, StyledStars } from '../../styled/other/StarBar';
import { colors } from '../../styled/globalStyled';

interface TypeProps {
  count: string
}

export const StarBar: React.FC<TypeProps> = props => {
    const stars: number[] = Array(5).fill(0);
    const { count } = props;
    return (
        <StyledStars>
          {stars.map((_: number, index: number) => (
              <StyledBsStarFill
                key={index}
                size={20}
                color={parseInt(count, 10) > index ? colors.starActive : colors.starDisable}
              />
            ))}
        </StyledStars>
    );
};
