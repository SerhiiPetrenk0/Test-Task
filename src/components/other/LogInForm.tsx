import React from 'react';
import { Form, Button, Modal } from 'react-bootstrap';
import { useDispatch  } from 'react-redux';
import { useFormik } from 'formik';

import { StyledValidEmail } from '../../styled/other/LogInForm';
import { fillUser } from '../../redux/ducks/userInfoDuck';
import { TypeFormikLogInForm, TypeUserinfo } from '../../interface';
import { YopLoginForm } from '../../validation';

interface TypeProps {
  show: boolean,
  handleClose: () => void
}

export const LogInForm: React.FC<TypeProps> = props => {
    const { show, handleClose } = props;
    const dispatch = useDispatch();
    const formik: TypeFormikLogInForm = useFormik({
      initialValues: {
        email: '',
        password: ''
      },
      validationSchema: YopLoginForm,
      onSubmit: (values: TypeUserinfo) => {
        dispatch(fillUser(formik.values))
        handleClose()
      }
    });
    
    const emailLog = !!formik.errors.email ?
      formik.errors.email :
      "We'll never share your email with anyone else.";
    const emailMassage: JSX.Element = (
      <StyledValidEmail check={formik.errors.email}>
        {emailLog}
      </StyledValidEmail>
    );

    const passwordLog = !!formik.errors.password ?
      formik.errors.password : 
      "Your password must be 6 to 16 characters long, contain letters, numbers, special characters !@#$%^&*, And no spaces or emojis."
    const passwordMassage: JSX.Element = (
      <StyledValidEmail check={formik.errors.password}>
        {passwordLog}
      </StyledValidEmail>
    );

    return (
      <>
        <Modal 
          show={show} 
          onHide={handleClose} 
          backdrop="static" 
          keyboard={false}
        >
          <Modal.Header closeButton>
            <Modal.Title>Registration</Modal.Title>
          </Modal.Header>
          <Modal.Body>
            <Form onSubmit={formik.handleSubmit}>
              <Form.Group className="mb-3" controlId="formBasicEmail">
                <Form.Label>Email address</Form.Label>
                <Form.Control 
                  name="email" 
                  onChange={formik.handleChange} 
                  value={formik.values.email} 
                  type="email" 
                  placeholder="Enter email" 
                />
                {emailMassage}
              </Form.Group>

              <Form.Group className="mb-3" controlId="formBasicPassword">
                <Form.Label>Password</Form.Label>
                <Form.Control 
                  name="password" 
                  onChange={formik.handleChange} 
                  value={formik.values.password} 
                  type="password" 
                  placeholder="Password" 
                />
                {passwordMassage}
              </Form.Group>

              <Button variant="primary" type="submit">
                Submit
              </Button>
            </Form>
          </Modal.Body>
        </Modal>
      </>
    );
} ;
