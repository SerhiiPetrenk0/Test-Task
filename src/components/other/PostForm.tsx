import React , { useState } from 'react';
import { Form , Button } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';

import { LogInForm } from './LogInForm';

import { formComment } from '../../redux/ducks/productsDuck';
import { colors } from '../../styled/globalStyled';
import { StyledContainer, StyledFormGroup, StyledBsStarFill } from '../../styled/other/PostForm';
import { TypeUserinfo, TypePostForm } from '../../interface';

export const PostForm:React.FC = () => {
    const [currentValue, setCurrentValue] = useState<number>(0);
    const [hoverValue, setHoverValue] = useState<undefined | number>(undefined);
    const [show, setShow] = useState<boolean>(false);
    const [formValue, setFormValue] = useState<TypePostForm>(
        {
            rating: '',
            comments: '',
        }
    );
    
    const stars: number[] = Array(5).fill(0);
    const dispatch = useDispatch();
    const userInfo: TypeUserinfo = useSelector((store:{userInfo:{userInfo: TypeUserinfo}}) => store.userInfo.userInfo);

    const handleMouseOver = (newHoverValue: React.SetStateAction<number | undefined>) => setHoverValue(newHoverValue);
    const handleMouseLeave = () => setHoverValue(undefined);
    const handleTextarea = (val: string) => setFormValue({...formValue, comments:val});
    const handleClose = () => setShow(false);
    const handleShow = () => setShow(true);
    const handleSubmit = (event: { preventDefault: () => void; }) => {
        event.preventDefault();
        if (!!userInfo.email) {
            dispatch(formComment(formValue));
            setFormValue({ ...formValue, comments:''});
            setCurrentValue(0);
        } else {
            handleShow()
        }
    };
    const handleClick = (value: React.SetStateAction<number>) => {
        setCurrentValue(value);
        setFormValue({...formValue, rating: value.toString()});
    };

    return (
        <>
            <LogInForm show={show} handleClose={handleClose} />
            <StyledContainer>
                <Form onSubmit={handleSubmit}>
                    <StyledFormGroup controlId="exampleForm.ControlInput1">
                        {stars.map((_, index) => (
                                <StyledBsStarFill
                                key={index}
                                size={24}
                                onClick={() => handleClick(index + 1)}
                                onMouseOver={() => handleMouseOver(index + 1)}
                                onMouseLeave={handleMouseLeave}
                                color={(hoverValue || currentValue) > index ? colors.starActive : colors.starDisable}
                                />
                        ))}
                    </StyledFormGroup>
                    <StyledFormGroup controlId="exampleForm.ControlTextarea1">
                        <Form.Label>comment</Form.Label>
                        <Form.Control as="textarea" rows={3} value={formValue.comments} onChange={event => handleTextarea(event.target.value)} />
                    </StyledFormGroup>
                    <Button variant="primary" type="submit">
                            Submit
                    </Button>
                </Form>
            </StyledContainer>
        </>
    );
};
