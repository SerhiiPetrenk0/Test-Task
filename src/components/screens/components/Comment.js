import React from 'react';
import { Card } from 'react-bootstrap';
import { StarBar } from './StarBar';
import { StyledCard } from '../../../styled/screens/components/Comment';

export const Comment = props => {
    const { ListComment } = props
    return(
        <StyledCard>
            <Card.Body>
                <Card.Text>
                {ListComment.discription}
                </Card.Text>
                <Card.Subtitle>
                <StarBar count={ListComment.stars} />
                </Card.Subtitle>
            </Card.Body>
        </StyledCard>
    )
}