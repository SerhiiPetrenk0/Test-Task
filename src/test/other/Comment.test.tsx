import React from "react";
import { configure, shallow } from "enzyme";
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJson from "enzyme-to-json";

import { Comment } from "../../components/other/Comment";

configure({ adapter: new Adapter() })

describe('Comment component', () => {
    const dataItem = {
        rating: "4",
        comments: "Odio ut qui. Autem cumque deleniti. Quia provident inventore. Enim ipsum totam."
    }    
    it('Snapshot Comment show', () => {
        const component = shallow(<Comment ListComment={dataItem} />)
        expect(toJson(component)).toMatchSnapshot();
    })
})