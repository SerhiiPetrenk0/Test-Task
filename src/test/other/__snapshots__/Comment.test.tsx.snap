// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment component Snapshot Comment show 1`] = `
<Styled(Card)>
  <CardBody>
    <CardText>
      Odio ut qui. Autem cumque deleniti. Quia provident inventore. Enim ipsum totam.
    </CardText>
    <CardSubtitle>
      <StarBar
        count="4"
      />
    </CardSubtitle>
  </CardBody>
</Styled(Card)>
`;
